### ISSUE TITLE
MAPQ does not follow SAM specifications when -k or -a is used.

### ISSUE THREAD

**Issue by d-cameron** [BenLangmead/bowtie2#78]
I have aligned reads using bowtie2 -k 100 and the MAPQs being reported are indeed not meaningful as per

```
 Reporting:
  (default)          look for multiple alignments, report best, with MAPQ
   OR
  -k <int>           report up to <int> alns per read; MAPQ not meaningful
   OR
  -a/--all           report all alignments; very slow, MAPQ not meaningful
```

The SAM specifications require the following:
```
1.4.5. MAPQ: MAPping Quality. It equals −10 log10 Pr{mapping position is wrong}, rounded to the nearest
integer. A value 255 indicates that the mapping quality is not available.
```

Can bowtie2 be updated to report MAPQ of 255 when used with the -k or -a flags, as reporting an incorrect MAPQ score causes serious issues with downstream variant calling. Alternatively, a MAPQ of 255 for multi-mapping reads and the default bowtie2 mapq score for uniquely mapping reads would be even better!

**Comment by BenLangmead** [comment]
What is the "incorrect" MAPQ score you're seeing?

**Comment by d-cameron** [comment]
Aligning the Illumina Platinum Genomics 50x NA12878 sample with -k 100 I get records such as this:

```
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        99      chrX    107118589       25      101M    =       107118765       277     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:194        XS:i:141        XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:32C68      YS:i:202        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        147     chrX    107118765       25      101M    =       107118589       -277    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:202        XS:i:123        XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:101        YS:i:194        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr10   81366184        25      2S49M1I7M3D42M  =       81366365        257     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:98 XS:i:141        XN:i:0  XM:i:10 XO:i:2  XG:i:4  NM:i:14 MD:Z:8T2G14A3C10T0G7C1T3^CAA33C3A4      YS:i:54 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr10   81366365        25      5S69M27S        =       81366184        -257    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:54 XS:i:123        XN:i:0  XM:i:11 XO:i:0  XG:i:0  NM:i:11 MD:Z:4C6A2G15G8T0G4C0T6A0T8G5   YS:i:98 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        419     chr10   108031154       25      6M1I94M =       108031328       273     GGACCACCATTCTGGTGGCCTCTGAACAAACTATTTGTCTCCCCCTAGTGAAGACTGATATTAACACAGAAGTTTGGGCAATTCAAGGGAAAATTGGCCAA   CCCFFFFFHHHHHJJEHHIJJJJJJIJJJJJIJJJJIGFHIJJJJGGIHIDFHIIJJJIJJJJJJJJJIGGHAHHHHFFFDDEECEDDDDDDDCCDDCDDD   AS:i:123        XS:i:123        XN:i:0  XM:i:9  XO:i:1  XG:i:1  NM:i:10 MD:Z:9C3A5C1A1G25G4C0A31T12     YS:i:100        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        339     chr10   108031328       25      42M3D4M1I50M4S  =       108031154       -273    AAACCAGAAGTTAGGAAAGGACTAGAAGCCATCATTGATAACAAGGTTGCAGGGCCTCCTCAAACCCTCCAACAGCCCTTGTAATACTCCGATATTGGGAG   <C>CDDDEDDDDDDDDDDCCADDCCDDDEEEDEEFFFFFFFHHHHHJJIIJJJJJIIIJIHGF?JJIHJIFGIGJIJJJJJJJHGEJIHHHHHFFFFFCCB   AS:i:100        XS:i:141        XN:i:0  XM:i:9  XO:i:2  XG:i:4  NM:i:13 MD:Z:12G29^TTG3A4A0C4G10A3T2T11C9       YS:i:123        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr11   56766796        25      2S51M2I1M1D3M3D42M      =       56766975        280     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:84 XS:i:141        XN:i:0  XM:i:10 XO:i:3  XG:i:6  NM:i:16 MD:Z:8T2G5G7C0A2T0C14G1T4^T3^CAA30G11   YS:i:60 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr11   56766975        25      3S41M1I50M1I5M  =       56766796        -280    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:60 XS:i:123        XN:i:0  XM:i:15 XO:i:2  XG:i:2  NM:i:17 MD:Z:13C2G12T2G0T4A2G4C9C7A6C0A3G5T3G9  YS:i:84 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        419     chr14   103633919       25      101M    =       103634415       597     GGACCACCATTCTGGTGGCCTCTGAACAAACTATTTGTCTCCCCCTAGTGAAGACTGATATTAACACAGAAGTTTGGGCAATTCAAGGGAAAATTGGCCAA   CCCFFFFFHHHHHJJEHHIJJJJJJIJJJJJIJJJJIGFHIJJJJGGIHIDFHIIJJJIJJJJJJJJJIGGHAHHHHFFFDDEECEDDDDDDDCCDDCDDD   AS:i:94 XS:i:123        XN:i:0  XM:i:14 XO:i:0  XG:i:0  NM:i:14 MD:Z:10C3A5C1A1G7C3A5T3T3G4C0A8C28G6    YS:i:141        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        339     chr14   103634415       25      42M2D57M2S      =       103633919       -597    AAACCAGAAGTTAGGAAAGGACTAGAAGCCATCATTGATAACAAGGTTGCAGGGCCTCCTCAAACCCTCCAACAGCCCTTGTAATACTCCGATATTGGGAG   <C>CDDDEDDDDDDDDDDCCADDCCDDDEEEDEEFFFFFFFHHHHHJJIIJJJJJIIIJIHGF?JJIHJIFGIGJIJJJJJJJHGEJIHHHHHFFFFFCCB   AS:i:141        XS:i:141        XN:i:0  XM:i:6  XO:i:1  XG:i:2  NM:i:8  MD:Z:14A27^TT23T2G13C4C2A8      YS:i:94 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr18   28209441        25      3S50M1I5M3D42M  =       28209621        222     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:98 XS:i:141        XN:i:0  XM:i:10 XO:i:2  XG:i:4  NM:i:14 MD:Z:10G3C10A3C12A8T3^CAA9A13A3T4T9     YS:i:60 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr18   28209621        25      5S34M62S        =       28209441        -222    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:60 XS:i:123        XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:30G3       YS:i:98 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        419     chr21   18554614        25      98M3S   =       18554789        272     GGACCACCATTCTGGTGGCCTCTGAACAAACTATTTGTCTCCCCCTAGTGAAGACTGATATTAACACAGAAGTTTGGGCAATTCAAGGGAAAATTGGCCAA   CCCFFFFFHHHHHJJEHHIJJJJJJIJJJJJIJJJJIGFHIJJJJGGIHIDFHIIJJJIJJJJJJJJJIGGHAHHHHFFFDDEECEDDDDDDDCCDDCDDD   AS:i:110        XS:i:123        XN:i:0  XM:i:11 XO:i:0  XG:i:0  NM:i:11 MD:Z:10C3A5A1A1G7C3C4T8G5A8C32  YS:i:118        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        339     chr21   18554789        25      42M2D38M1I15M5S =       18554614        -272    AAACCAGAAGTTAGGAAAGGACTAGAAGCCATCATTGATAACAAGGTTGCAGGGCCTCCTCAAACCCTCCAACAGCCCTTGTAATACTCCGATATTGGGAG   <C>CDDDEDDDDDDDDDDCCADDCCDDDEEEDEEFFFFFFFHHHHHJJIIJJJJJIIIJIHGF?JJIHJIFGIGJIJJJJJJJHGEJIHHHHHFFFFFCCB   AS:i:118        XS:i:141        XN:i:0  XM:i:7  XO:i:2  XG:i:3  NM:i:10 MD:Z:13A18T9^TT9A16G0G16C2A5    YS:i:110        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        419     chr3    149009122       25      81M1D5M1I11M3S  =       149009298       277     GGACCACCATTCTGGTGGCCTCTGAACAAACTATTTGTCTCCCCCTAGTGAAGACTGATATTAACACAGAAGTTTGGGCAATTCAAGGGAAAATTGGCCAA   CCCFFFFFHHHHHJJEHHIJJJJJJIJJJJJIJJJJIGFHIJJJJGGIHIDFHIIJJJIJJJJJJJJJIGGHAHHHHFFFDDEECEDDDDDDDCCDDCDDD   AS:i:94 XS:i:123        XN:i:0  XM:i:11 XO:i:2  XG:i:2  NM:i:13 MD:Z:10C3A5C3G0G6C0G16G14T15^C2G6G6     YS:i:113        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        339     chr3    149009298       25      42M3D4M1I52M2S  =       149009122       -277    AAACCAGAAGTTAGGAAAGGACTAGAAGCCATCATTGATAACAAGGTTGCAGGGCCTCCTCAAACCCTCCAACAGCCCTTGTAATACTCCGATATTGGGAG   <C>CDDDEDDDDDDDDDDCCADDCCDDDEEEDEEFFFFFFFHHHHHJJIIJJJJJIIIJIHGF?JJIHJIFGIGJIJJJJJJJHGEJIHHHHHFFFFFCCB   AS:i:113        XS:i:141        XN:i:0  XM:i:8  XO:i:2  XG:i:4  NM:i:12 MD:Z:39G2^TTG3A2T2A15G7T10C2A8  YS:i:94 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr4    32090007        25      2S11M1I44M2D43M =       32090177        237     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:138        XS:i:141        XN:i:0  XM:i:5  XO:i:2  XG:i:3  NM:i:8  MD:Z:9A13G5C18A2T3^CG43 YS:i:68 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr4    32090177        25      1S11M1I53M35S   =       32090007        -237    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:68 XS:i:123        XN:i:0  XM:i:7  XO:i:1  XG:i:1  NM:i:8  MD:Z:3T13G15G8T3A1A9A5  YS:i:138        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr5    32665945        25      2S51M1I5M3D42M  =       32666121        279     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:107        XS:i:141        XN:i:0  XM:i:9  XO:i:2  XG:i:4  NM:i:13 MD:Z:4G4A1G9A4A3C5G15T3^CAA33C8 YS:i:94 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr5    32666121        25      101M    =       32665945        -279    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:94 XS:i:123        XN:i:0  XM:i:14 XO:i:0  XG:i:0  NM:i:14 MD:Z:35G8T0G4C9A2G0A3G7C1T1G5T2T0G10    YS:i:107        YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        355     chr6    119719112       25      2S9M1D3M1I44M2D42M      =       119719289       214     CTCCCAATATCGGAGTATTACAAGGGCTGTTGGAGGGTTTGAGGAGGCCCTGCAACCTTGTTATCAATGATGGCTTCTAGTCCTTTCCTAACTTCTGGTTT   BCCFFFFFHHHHHIJEGHJJJJJJJIJGIGFIJHIJJ?FGHIJIIIJJJJJIIJJHHHHHFFFFFFFEEDEEEDDDCCDDACCDDDDDDDDDDEDDDC>C<   AS:i:101        XS:i:141        XN:i:0  XM:i:9  XO:i:3  XG:i:4  NM:i:13 MD:Z:9^A10G5A3C1A14T9^AA3G14C1A11T9     YS:i:55 YT:Z:CP
HSQ1004:134:C0D8DACXX:1:1101:3898:130991        403     chr6    119719289       25      4S31M66S        =       119719112       -214    TTGGCCAATTTTCCCTTGAATTGCCCAAACTTCTGTGTTAATATCAGTCTTCACTAGGGGGAGACAAATAGTTTGTTCAGAGGCCACCAGAATGGTGGTCC   DDDCDDCCDDDDDDDECEEDDFFFHHHHAHGGIJJJJJJJJJIJJJIIHFDIHIGGJJJJIHFGIJJJJIJJJJJIJJJJJJIHHEJJHHHHHFFFFFCCC   AS:i:55 XS:i:123        XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:7C23       YS:i:101        YT:Z:CP

```

The reads with only 1 mapping location report a MAPQ of 255, but the multi-mapped reads report MAPQs whose probabilities sum to greater than 1. As far as I can tell, all alignments for a given read pair all report the same MAPQ. The top 10,000,000 (name sorted) records give a MAPQ distribution of:

```
  96614 0
 717109 1
 744583 2
  18131 9
1703673 11
   4106 12
 350425 14
   1080 16
 322849 17
 380856 18
     42 19
 419288 21
 383888 25
 400973 31
 409197 32
 503343 33
 662044 34
 629286 35
 996807 37
 622099 38
 197924 39
  23130 40
 412553 255
```

I would have expected a preponderance of records with MAPQ  in the 0-2 range for the multi-mapped reads. Unfortunately, the high MAPQ scores for reads with multiple mapping locations causes issues with downstream variant caller who expect the reported MAPQ to be approximately correct.

**Comment by ch4rr0** [comment]
> Can bowtie2 be updated to report MAPQ of 255 when used with the -k or -a flags

This change is available in v2.4.0 and later.
